File: android/src/main/java/com/reactlibrary/TenjinModule.java
// TenjinModule.java

package com.reactlibrary;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.tenjin.android.TenjinSDK;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class TenjinModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;
    public TenjinSDK instance = null;

    public TenjinModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "Tenjin";
    }

    @ReactMethod
    public void initialize(String apiKey) {
        instance = TenjinSDK.getInstance(reactContext, apiKey);
    }

    @ReactMethod
    public void connect() {
        instance.connect();
    }

    @ReactMethod
    public void setAppStore(String type) {
        TenjinSDK.AppStoreType appStoreType = TenjinSDK.AppStoreType.googleplay;
        switch (type) {
            case "googleplay":
                appStoreType = TenjinSDK.AppStoreType.googleplay;
                break;
            case "amazon":
                appStoreType = TenjinSDK.AppStoreType.amazon;
                break;
            case "other":
                appStoreType = TenjinSDK.AppStoreType.other;
                break;
            default:
                appStoreType = TenjinSDK.AppStoreType.unspecified;
                break;
        }
        instance.setAppStore(appStoreType);
    }

    @ReactMethod
    public void optIn() {
        instance.optIn();
    }

    @ReactMethod
    public void optOut() {
        instance.optOut();
    }

    @ReactMethod
    public void optInParams(ReadableArray params) {
        instance.optInParams(readableToArray(params));
    }

    @ReactMethod
    public void optOutParams(ReadableArray params) {
        instance.optOutParams(readableToArray(params));
    }

    @ReactMethod
    public void transaction(String productName, String currencyCode, Double quantity, Double unitPrice) {
        instance.transaction(productName, currencyCode, Integer.valueOf(quantity.intValue()), unitPrice);
    }

    @ReactMethod
    public void eventWithName(String name) {
        instance.eventWithName(name);
    }

    @ReactMethod
    public void eventWithNameAndValue(String name, String value) {
        instance.eventWithNameAndValue(name, value);
    }

    @ReactMethod
    public void appendAppSubversion(Double version) {
        instance.appendAppSubversion(Integer.valueOf(version.intValue()));
    }

    @ReactMethod
    public void getAttributionInfo(com.facebook.react.bridge.Callback successCallback, com.facebook.react.bridge.Callback errorCallback) {
       instance.getAttributionInfo(data -> {
           try {
               successCallback.invoke(convertJsonToMap(new JSONObject(data)));
           } catch (JSONException e) {
               e.printStackTrace();
           }
       });
    }

    @ReactMethod
    public void eventAdImpressionAdMob(ReadableMap json) {
        try {
            instance.eventAdImpressionAdMob(convertMapToJson(json));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @ReactMethod
    public void eventAdImpressionAppLovin(ReadableMap json) {
        try {
            instance.eventAdImpressionAppLovin(convertMapToJson(json));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @ReactMethod
    public void eventAdImpressionHyperBid(ReadableMap json) {
        try {
            instance.eventAdImpressionHyperBid(convertMapToJson(json));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @ReactMethod
    public void eventAdImpressionIronSource(ReadableMap json) {
        try {
            instance.eventAdImpressionIronSource(convertMapToJson(json));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @ReactMethod
    public void eventAdImpressionTopOn(ReadableMap json) {
        try {
            instance.eventAdImpressionTopOn(convertMapToJson(json));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private String[] readableToArray(ReadableArray readableArray) {
        String[] array = new String[readableArray.size()];
        for (int i = 0; i < readableArray.size(); i++) {
            array[i] = readableArray.getString(i);
        }
        return array;
    }

    private JSONObject convertMapToJson(ReadableMap readableMap) throws JSONException {
        JSONObject object = new JSONObject();
        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
        while (iterator.hasNextKey()) {
            String key = iterator.nextKey();
            switch (readableMap.getType(key)) {
                case Null:
                    object.put(key, JSONObject.NULL);
                    break;
                case Boolean:
                    object.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    object.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    object.put(key, readableMap.getString(key));
                    break;
                case Map:
                    object.put(key, convertMapToJson(readableMap.getMap(key)));
                    break;
                case Array:
                    object.put(key, convertArrayToJson(readableMap.getArray(key)));
                    break;
            }
        }
        return object;
    }

    private JSONArray convertArrayToJson(ReadableArray readableArray) throws JSONException {
        JSONArray array = new JSONArray();
        for (int i = 0; i < readableArray.size(); i++) {
            switch (readableArray.getType(i)) {
                case Null:
                    break;
                case Boolean:
                    array.put(readableArray.getBoolean(i));
                    break;
                case Number:
                    array.put(readableArray.getDouble(i));
                    break;
                case String:
                    array.put(readableArray.getString(i));
                    break;
                case Map:
                    array.put(convertMapToJson(readableArray.getMap(i)));
                    break;
                case Array:
                    array.put(convertArrayToJson(readableArray.getArray(i)));
                    break;
            }
        }
        return array;
    }

    public WritableMap convertJsonToMap(JSONObject jsonObject) throws JSONException {
        WritableMap map = new WritableNativeMap();

        Iterator<String> iterator = jsonObject.keys();
        while (iterator.hasNext()) {
            String key = iterator.next();
            Object value = jsonObject.get(key);
            if (value instanceof JSONObject) {
                map.putMap(key, convertJsonToMap((JSONObject) value));
            } else if (value instanceof JSONArray) {
                map.putArray(key, convertJsonToArray((JSONArray) value));
                if (("option_values").equals(key)) {
                    map.putArray("options", convertJsonToArray((JSONArray) value));
                }
            } else if (value instanceof Boolean) {
                map.putBoolean(key, (Boolean) value);
            } else if (value instanceof Integer) {
                map.putInt(key, (Integer) value);
            } else if (value instanceof Double) {
                map.putDouble(key, (Double) value);
            } else if (value instanceof String) {
                map.putString(key, (String) value);
            } else {
                map.putString(key, value.toString());
            }
        }
        return map;
    }

    public WritableArray convertJsonToArray(JSONArray jsonArray) throws JSONException {
        WritableArray array = new WritableNativeArray();

        for (int i = 0; i < jsonArray.length(); i++) {
            Object value = jsonArray.get(i);
            if (value instanceof JSONObject) {
                array.pushMap(this.convertJsonToMap((JSONObject) value));
            } else if (value instanceof JSONArray) {
                array.pushArray(convertJsonToArray((JSONArray) value));
            } else if (value instanceof Boolean) {
                array.pushBoolean((Boolean) value);
            } else if (value instanceof Integer) {
                array.pushInt((Integer) value);
            } else if (value instanceof Double) {
                array.pushDouble((Double) value);
            } else if (value instanceof String) {
                array.pushString((String) value);
            } else {
                array.pushString(value.toString());
            }
        }
        return array;
    }
}

File: android/src/main/java/com/tenjin/TenjinModule.kt
package com.tenjin

import com.facebook.react.bridge.*
import com.tenjin.android.TenjinSDK
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean

class TenjinModule internal constructor(private val reactContext: ReactApplicationContext): NativeTenjinSpec(reactContext)  {
  companion object {
    const val NAME = "Tenjin"
  }

  override fun getName(): String = NAME
  private var instance: TenjinSDK? = null

  @ReactMethod
  override fun initialize(apiKey: String) {
    instance = TenjinSDK.getInstance(reactContext, apiKey)
  }

  @ReactMethod
  override fun connect() {
    instance?.connect()
  }

  @ReactMethod
  override fun setAppStore(type: String) {
    val appStoreType = when (type) {
      "googleplay" -> TenjinSDK.AppStoreType.googleplay
      "amazon" -> TenjinSDK.AppStoreType.amazon
      "other" -> TenjinSDK.AppStoreType.other
      else -> TenjinSDK.AppStoreType.unspecified
    }
    instance?.setAppStore(appStoreType)
  }

  @ReactMethod
  override fun optIn() {
    instance?.optIn()
  }

  @ReactMethod
  override fun optOut() {
    instance?.optOut()
  }

  @ReactMethod
  override fun optInParams(params: ReadableArray) {
    instance?.optInParams(readableToArray(params))
  }

  @ReactMethod
  override fun optOutParams(params: ReadableArray) {
    instance?.optOutParams(readableToArray(params))
  }

  @ReactMethod
  override fun optInOutUsingCMP() {
    instance?.optInOutUsingCMP()
  }

  @ReactMethod
  override fun optOutGoogleDMA() {
    instance?.optOutGoogleDMA()
  }

  @ReactMethod
  override fun optInGoogleDMA() {
    instance?.optInGoogleDMA()
  }

  @ReactMethod
  override fun transaction(productName: String, currencyCode: String, quantity: Double, unitPrice: Double) {
    instance?.transaction(productName, currencyCode, quantity.toInt(), unitPrice)
  }

  @ReactMethod
  override fun transactionWithReceipt(
    productName: String,
    currencyCode: String,
    quantity: Double,
    unitPrice: Double,
    transaction: String,
    data: String
  ) {
    instance?.transaction(productName, currencyCode, quantity.toInt(), unitPrice, transaction, data)
  }

  @ReactMethod
  override fun transactionWithDataSignature(
    productName: String,
    currencyCode: String,
    quantity: Double,
    unitPrice: Double,
    purchaseData: String,
    dataSignature: String
  ) {
    instance?.transaction(productName, currencyCode, quantity.toInt(), unitPrice, purchaseData, dataSignature)
  }

  @ReactMethod
  override fun eventWithName(name: String) {
    instance?.eventWithName(name)
  }

  @ReactMethod
  override fun eventWithNameAndValue(name: String, value: String) {
    instance?.eventWithNameAndValue(name, value)
  }

  @ReactMethod
  override fun appendAppSubversion(version: Double) {
    instance?.appendAppSubversion(version.toInt())
  }

  @ReactMethod
  override fun getAttributionInfo(successCallback: Callback, errorCallback: Callback) {
    val callbackInvoked = AtomicBoolean(false)
    instance?.getAttributionInfo { data ->
      if (!callbackInvoked.getAndSet(true)) {
        try {
          successCallback.invoke(convertJsonToMap(JSONObject(data as Map<String, Any>)))
        } catch (e: JSONException) {
          errorCallback.invoke("JSON Parsing Error: ${e.message}")
        }
      }
    }
  }

  @ReactMethod
  override fun eventAdImpressionAdMob(json: ReadableMap) {
    try {
      instance?.eventAdImpressionAdMob(convertMapToJson(json))
    } catch (e: JSONException) {
      e.printStackTrace()
    }
  }

  @ReactMethod
  override fun eventAdImpressionAppLovin(json: ReadableMap) {
    try {
      instance?.eventAdImpressionAppLovin(convertMapToJson(json))
    } catch (e: JSONException) {
      e.printStackTrace()
    }
  }

  @ReactMethod
  override fun eventAdImpressionHyperBid(json: ReadableMap) {
    try {
      instance?.eventAdImpressionHyperBid(convertMapToJson(json))
    } catch (e: JSONException) {
      e.printStackTrace()
    }
  }

  @ReactMethod
  override fun eventAdImpressionIronSource(json: ReadableMap) {
    try {
      instance?.eventAdImpressionIronSource(convertMapToJson(json))
    } catch (e: JSONException) {
      e.printStackTrace()
    }
  }

  @ReactMethod
  override fun eventAdImpressionTopOn(json: ReadableMap) {
    try {
      instance?.eventAdImpressionTopOn(convertMapToJson(json))
    } catch (e: JSONException) {
      e.printStackTrace()
    }
  }

  @ReactMethod
  override fun eventAdImpressionTradPlus(json: ReadableMap) {
    try {
      instance?.eventAdImpressionTradPlus(convertMapToJson(json))
    } catch (e: JSONException) {
      e.printStackTrace()
    }
  }

  @ReactMethod
  override fun setCustomerUserId(userId: String) {
    instance?.setCustomerUserId(userId)
  }

  @ReactMethod
  override fun getCustomerUserId(callback: Callback) {
    callback.invoke(instance?.customerUserId ?: "")
  }

  @ReactMethod
  override fun getAnalyticsInstallationId(callback: Callback) {
    callback.invoke(instance?.analyticsInstallationId ?: "")
  }

  @ReactMethod
  override fun setGoogleDMAParameters(adPersonalization: Boolean, adUserData: Boolean) {
    instance?.setGoogleDMAParameters(adPersonalization, adUserData)
  }

  @ReactMethod
  override fun setCacheEventSetting(setting: Boolean) {
    instance?.setCacheEventSetting(setting)
  }

  @ReactMethod
  override fun setEncryptRequestsSetting(setting: Boolean) {
    instance?.setEncryptRequestsSetting(setting)
  }

  @ReactMethod
  override fun updatePostbackConversionValue(conversionValue: Double) {
    // Nothing to implement
  }

  @ReactMethod
  override fun updatePostbackConversionValueWithCoarseValue(conversionValue: Double, coarseValue: String) {
    // Nothing to implement
  }

  @ReactMethod
  override fun updatePostbackConversionValueWithCoarseValueAndLockWindow(conversionValue: Double, coarseValue: String, lockWindow: Boolean) {
    // Nothing to implement
  }

  @ReactMethod
  override fun getUserProfileDictionary(callback: Callback) {
    val profileDict = instance?.getUserProfileDictionary()
    if (profileDict != null) {
      try {
        val writableMap = convertMapToWritableMap(profileDict)
        callback.invoke(writableMap)
      } catch (e: Exception) {
        callback.invoke(WritableNativeMap())
      }
    } else {
      callback.invoke(WritableNativeMap())
    }
  }

  @ReactMethod
  override fun resetUserProfile() {
    instance?.resetUserProfile()
  }

  private fun convertMapToWritableMap(map: Map<String, Any>): WritableMap {
    val writableMap = WritableNativeMap()
    for ((key, value) in map) {
      when (value) {
        is Map<*, *> -> writableMap.putMap(key, convertMapToWritableMap(value as Map<String, Any>))
        is List<*> -> writableMap.putArray(key, convertListToWritableArray(value))
        is Boolean -> writableMap.putBoolean(key, value)
        is Int -> writableMap.putInt(key, value)
        is Long -> writableMap.putDouble(key, value.toDouble())
        is Double -> writableMap.putDouble(key, value)
        is String -> writableMap.putString(key, value)
        else -> writableMap.putString(key, value.toString())
      }
    }
    return writableMap
  }

  private fun convertListToWritableArray(list: List<*>): WritableArray {
    val writableArray = WritableNativeArray()
    for (value in list) {
      when (value) {
        is Map<*, *> -> writableArray.pushMap(convertMapToWritableMap(value as Map<String, Any>))
        is List<*> -> writableArray.pushArray(convertListToWritableArray(value))
        is Boolean -> writableArray.pushBoolean(value)
        is Int -> writableArray.pushInt(value)
        is Long -> writableArray.pushDouble(value.toDouble())
        is Double -> writableArray.pushDouble(value)
        is String -> writableArray.pushString(value)
        else -> writableArray.pushString(value.toString())
      }
    }
    return writableArray
  }

  private fun readableToArray(readableArray: ReadableArray): Array<String> {
    return Array(readableArray.size()) { readableArray.getString(it) ?: "" }
  }

  private fun convertMapToJson(readableMap: ReadableMap): JSONObject {
    val jsonObject = JSONObject()
    val iterator = readableMap.keySetIterator()
    while (iterator.hasNextKey()) {
      val key = iterator.nextKey()
      when (readableMap.getType(key)) {
        ReadableType.Null -> jsonObject.put(key, JSONObject.NULL)
        ReadableType.Boolean -> jsonObject.put(key, readableMap.getBoolean(key))
        ReadableType.Number -> jsonObject.put(key, readableMap.getDouble(key))
        ReadableType.String -> jsonObject.put(key, readableMap.getString(key))
        ReadableType.Map -> jsonObject.put(key, convertMapToJson(readableMap.getMap(key)!!))
        ReadableType.Array -> jsonObject.put(key, convertArrayToJson(readableMap.getArray(key)!!))
      }
    }
    return jsonObject
  }

  private fun convertArrayToJson(readableArray: ReadableArray): JSONArray {
    val jsonArray = JSONArray()
    for (i in 0 until readableArray.size()) {
      when (readableArray.getType(i)) {
        ReadableType.Null -> jsonArray.put(JSONObject.NULL)
        ReadableType.Boolean -> jsonArray.put(readableArray.getBoolean(i))
        ReadableType.Number -> jsonArray.put(readableArray.getDouble(i))
        ReadableType.String -> jsonArray.put(readableArray.getString(i))
        ReadableType.Map -> {
          readableArray.getMap(i)?.let { map ->
            jsonArray.put(convertMapToJson(map))
          } ?: jsonArray.put(JSONObject.NULL)
        }
        ReadableType.Array -> {
          readableArray.getArray(i)?.let { array ->
            jsonArray.put(convertArrayToJson(array))
          } ?: jsonArray.put(JSONObject.NULL)
        }
      }
    }
    return jsonArray
  }

  private fun convertJsonToMap(jsonObject: JSONObject): WritableMap {
    val map = WritableNativeMap()
    val iterator = jsonObject.keys()
    while (iterator.hasNext()) {
      val key = iterator.next()
      val value = jsonObject[key]
      when (value) {
        is JSONObject -> map.putMap(key, convertJsonToMap(value))
        is JSONArray -> map.putArray(key, convertJsonToArray(value))
        is Boolean -> map.putBoolean(key, value)
        is Int -> map.putInt(key, value)
        is Double -> map.putDouble(key, value)
        is String -> map.putString(key, value)
        else -> map.putString(key, value.toString())
      }
    }
    return map
  }

  private fun convertJsonToArray(jsonArray: JSONArray): WritableArray {
    val array = WritableNativeArray()
    for (i in 0 until jsonArray.length()) {
      when (val value = jsonArray[i]) {
        is JSONObject -> array.pushMap(convertJsonToMap(value))
        is JSONArray -> array.pushArray(convertJsonToArray(value))
        is Boolean -> array.pushBoolean(value)
        is Int -> array.pushInt(value)
        is Double -> array.pushDouble(value)
        is String -> array.pushString(value)
        else -> array.pushString(value.toString())
      }
    }
    return array
  }
}

